Real cluster option and labelling (#93)
kyma-project/btp-manager: docs/developer/testing.md

## Running unit tests using CLI

To run unit tests from command line use following command (being in btp-manager main directory): 

```shell
make test
```
Details concerning `test` rule (prerequisites and recipe) are defined in `./Makefile`.

By default, unit tests are run using `env-test` cluster. 
Part of unit tests is implemented using [gingko](https://onsi.github.io/ginkgo/#top) library but all unit tests are invoked using `go test ./... <some options`
You can find exact invocation reflected in the console output along with messages confirming whether `env-test` is being used.
```
STEP: bootstrapping test environment @ 01/13/23 08:24:45.981
2023-01-13T08:24:45.981+0100  DEBUG   controller-runtime.test-env     starting control plane
```

### Running unit tests on existing cluster

You can run tests on existing cluster (not `env-test`) executing helper script:

```shell
testing/run-unit-tests-on-cluster.sh
```

Script sets environment variables accordingly and invokes the same `make` target - i.e. `test`. 

> Ginkgo unit tests will use cluster in the current context, name of the cluster can be found in console output e.g.:

Reviewer: ```suggestion
Ginkgo unit tests use the cluster in the current context. The name of the cluster can be found in the console output, for example:
```

```
Running unit tests on the existing "k3d-test-cluster" cluster
```
> **NOTE:** Script attempts to restore the proper state of the cluster before each test run by removing resources that could be left if tests failed to properly remove those.
> If problems with left-over resources recur, you can recreate the cluster.

### Test output verbosity

Setting for the `go test` verbosity is `-v` (verbose, print the full output event for passing tests). This can be changed in `make` recipe. 
For `ginkgo` tests execution default setting is `-ginkgo.v` (verbose) and can be changed e.g. for `very verbose` by setting environment variable:
```shell

> export GINKGO_VERBOSE_FLAG=-ginkgo.vv

Reviewer 2: put this GINKGO_VERBOSE_FLAG env variable to the script with default value, add a comment with very verbose value too.

Author: Done.

Author: Pushed.

testing/run-unit-tests-on-cluster.sh
```
Alternatively it can be changed in `./Makefile` e.g.
```Makefile

> GINKGO_VERBOSE_FLAG ?= -ginkgo.vv

Reviewer 2: when moved to set-env-vars.sh this is not needed. All testing envs will be placed in the set-env-test.sh

Author: Changes pushed.###

```

### Filtering labels
You can use `gingko` library labeling features to filter which tests specs are to be executed 
(for details see [Spec Labels](https://onsi.github.io/ginkgo/#spec-labels) in ginkgo documentation). In order to use labels for filtering 
you need to instrument test nodes (`Describe`, `It`, `When` et al.) in `./controllers/btpoperator_controller_test.go` with labels e.g.:
```go
	Describe("Provisioning", Label("test-provisioning", "smoke-test"), func() {
```
```go
	Describe("Deprovisioning", Label("test-deprovisioning"), func() {
```

> You can use labels either by redefining TEST_LABEL variable in `run-unit-tests-on-cluster.sh` for example to run only specs 
> labeled as `smoke-test`:
> ```shell
> TEST_LABELS="smoke-test"
> ```
> Or you can invoke script with parameter, for example:
> ```shell
> testing/run-unit-tests-on-cluster.sh "smoke-test"
> ```

Reviewer: ```suggestion
You can use labels by:
-  redefining **TEST_LABEL** variable in `run-unit-tests-on-cluster.sh`, for example, to run only specs labeled as `smoke-test`:
```shell
TEST_LABELS="smoke-test"
```
- invoking a script with a parameter, for example:
```shell
testing/run-unit-tests-on-cluster.sh "smoke-test"
```
```
